### PR TITLE
Move to the New Interface

### PR DESCRIPTION
**Pull request by arnavs**
First pass at the new interface. 

It was difficult for me to redesign this with the method hierarchy from before (things calling `get_operator_basis`, etc.), so these methods are pretty self-contained. That leads to a fair amount of redundant code. I'll work with @chiyahn to fix that.

### CODE REVIEW

**Comment by arnavs** [comment]
The key file to look at is the `src/types.jl`, and maybe the exports/tests. The `src/operators.jl` is basically a bunch of:

```
function DifferentialOperator(x::AbstractRange, bc::Tuple{Reflecting, Reflecting}, method::ForwardFirstDifference)
```

File: src/types.jl [hunk 1]
```diff
@@ -1,15 +1,21 @@
...
+    # Abstracts
+    abstract type BoundaryCondition end
+    abstract type HomogeneousBoundaryCondition <: BoundaryCondition end
+    abstract type InhomogeneousBoundaryCondition <: BoundaryCondition end
+    # Concretes
+    struct Reflecting <: HomogeneousBoundaryCondition end
+    struct Mixed{T} <: HomogeneousBoundaryCondition where {T <: Real}
+        ξ::T
+    end
+    struct Absorbing{T1, T2} <: InhomogeneousBoundaryCondition where {T1 <: Real, T2 <: Real}
```
Comment (jlperla) [inline src/types.jl]: Can remove this until we are ready for it.

Comment (arnavs) [reply src/types.jl]: Done.

**Comment by arnavs** [comment]
@chiyahn Assigning this to you. Merge whenever you feel ready to move to the new interface. 

If you could add in documentation as you feel appropriate, that would be a big help too.

**Comment by chiyahn** [comment]
Sounds good

**Comment by chiyahn** [comment]
@arnavs @jlperla Should be ready to go -- merge it when Travis CI passes

**Pull request merged**